design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,135 @@
+# Restore API Group Version by Priority Level When EnableAPIGroupVersions Feature is Set
+
+Status: Draft
+
+## Abstract
+
+This document proposes a solution to select an API group version to restore from the versions backed up using the feature EnableAPIGroupVersions.
+
+## Background
+
+It is possible that between the time a backup has been made and a restore occurs that the target Kubernetes version has incremented more than one version. In such a case where multiple versions of Kubernetes were skipped, the preferred source cluster's API group versions for resources may no longer be supported by the target cluster. With PR #2373, all supported API group versions were backed up if the EnableAPIGroupVersions feature flag was set during Velero installation. The next step (outlined by this design proposal) will be to see if any of the backed up versions are supported in the target cluster and if so, choose one to restore for each backed up resource.
+
+## Goals
+
+- Restore resources using the chosen API group version.
+
+## Non Goals
+
+- Allowing users to customize the predetermined API group versions priorities.
+- Modifying the compressed backup tarball files.
+- Using plugins to restore a resource when the target supports non of the source cluster's API group versions.
+
+## High-Level Design
+
+During restore, the proposal is that Velero will read the directory names in the backup tarball file and see which API group versions were backed up from the source cluster. Velero will also look at the target cluster and determine which API group versions are supported. Given the two lists of source and target cluster supported API versions, a priority system will be used to determine which version will be backed up. The priority of versions in order are target cluster preferred version, source cluster preferred version, and the latest common supported version. Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.
+
+## Detailed Design
+
+There are four main objectives to achieve the above stated goal:
+
+1. List the backed up API group versions.
```

Comment:
I believe it will handle the backward-compatible on the source cluster, which means it will support both _without_ and _with_ enabling feature flag `EnableAPIGroupVersions` on the source cluster backed up tarball.

Or, should it only compatible with format version >= 1.1.0?
https://github.com/vmware-tanzu/velero/blob/4178d9d/pkg/apis/velero/v1/backup.go#L233-L235

Reply:
> it will support both without and with enabling feature flag

I'm not sure what you mean. Without the feature flag, only the `source preferred version` will be in the backup tarball. No other supported versions will be there so I can only use the source preferred version when the feature flag is not enabled.

As for format version, the new format version applies to the all backup tarballs, with or without the feature flag. But without the feature flag, only to the source preferred version will use format version 1.1.0 because no other version will be in the tarball.

Only if the feature flag is enabled will the other supported versions be available and these versions will meet format version 1.1.0.

Let me know if I misunderstood you.

Reply:
I'm not sure if this will help. Here is what's in the backed tarball files. This first screenshot shows the autoscaler versions that are backed up without feature flag enabled:

![Screen Shot 2020-11-03 at 8 09 30 AM](https://user-images.githubusercontent.com/17328443/98010982-2d81e500-1dac-11eb-866a-ceb8a4a5c857.png)

This second screenshot shows the autoscaler versions that are backed up with the feature flag enabled:
![Screen Shot 2020-11-03 at 8 09 09 AM](https://user-images.githubusercontent.com/17328443/98011037-3d012e00-1dac-11eb-857f-1c879e279e84.png)

You can see that in both cases, format version 1.1.0 is being used, but the difference is how many versions were backed up.

Reply:
Both the screenshots if I remember correctly are commits format version `1.1.0`, it means that the source cluster and target cluster are running Velero after 1.4.

My question is will we support restores from the tarball with the legacy format version (Velero < 1.4)?
I guess yes because without enable the feature flag `EnableAPIGroupVersions`, the format version `1.1` and `1.1.0` almost look the same. Ref to https://github.com/vmware-tanzu/velero/blob/main/site/content/docs/main/output-file-format.md

Reply:
You bring up a really good point. Thank you for helping me to realize that what I proposed is not backward compatible with Velero versions earlier than 1.4.

If the backup tarball files do not follow `format version 1.1.0`, the code I have proposed will error when it shouldn't. It will try to set the chosen version to `v1-preferredversion`, but that version will not exist, and the restore will fail.

I'll update the design doc to check the directory names. If there is a directory with the substring `-preferredversion`, I know it is using `format version 1.1.0`. If not, I will let the program go on as usual, as if feature flag was not enabled.

Very good find, @jenting.